src/EnergyPlus/UserDefinedComponents.cc

```diff
@@ -128,30 +101,109 @@ namespace UserDefinedComponents {
-    bool GetInput(true);
 
-    // SUBROUTINE SPECIFICATIONS FOR MODULE <module_name>:
+    bool GetInput(true);
+    bool GetPlantCompInput(true);
 
-    // Object Data
     Array1D<UserPlantComponentStruct> UserPlantComp;
     Array1D<UserCoilComponentStruct> UserCoil;
     Array1D<UserZoneHVACForcedAirComponentStruct> UserZoneAirHVAC;
     Array1D<UserAirTerminalComponentStruct> UserAirTerminal;
 
-    // Functions
-
-    void SimUserDefinedPlantComponent(int const LoopNum,                       // plant loop sim call originated from
-                                      int const LoopSideNum,                   // plant loop side sim call originated from
-                                      std::string const &EP_UNUSED(EquipType), // type of equipment, 'PlantComponent:UserDefined'
-                                      std::string const &EquipName,            // user name for component
-                                      int &CompIndex,
-                                      bool &InitLoopEquip,
-                                      Real64 const MyLoad,
-                                      Real64 &MaxCap,
-                                      Real64 &MinCap,
-                                      Real64 &OptCap)
+    void clear_state()
+    {
+        NumUserPlantComps = 0;
+        NumUserCoils = 0;
+        NumUserZoneAir = 0;
+        NumUserAirTerminals = 0;
+
+        CheckUserPlantCompName.deallocate();
+        CheckUserCoilName.deallocate();
+        CheckUserZoneAirName.deallocate();
+        CheckUserAirTerminal.deallocate();
+
+        UserPlantComp.deallocate();
+        UserCoil.deallocate();
+        UserZoneAirHVAC.deallocate();
+        UserAirTerminal.deallocate();
+    }
+
+    PlantComponent *UserPlantComponentStruct::factory(std::string const &objectName)
+    {
+        // Process the input data
+        if (GetPlantCompInput) {
+            GetUserDefinedPlantComponents();
+            GetPlantCompInput = false;
+        }
+        // Now look for this particular object
+        for (auto &thisComp : UserPlantComp) {
+            if (thisComp.Name == objectName) {
+                return &thisComp;
+            }
+        }
+        // If we didn't find it, fatal
+        ShowFatalError("LocalUserDefinedPlantComponentFactory: Error getting inputs for object named: " + objectName); // LCOV_EXCL_LINE
+        // Shut up the compiler
+        return nullptr; // LCOV_EXCL_LINE
+    }
+
+    void UserPlantComponentStruct::onInitLoopEquip(const PlantLocation &calledFromLocation)
+    {
+        bool anyEMSRan;
+        Real64 myLoad = 0.0;
+        int thisLoop = 0;
+
+        this->initialize(calledFromLocation.loopNum, myLoad);
+
+        for (int loop = 1; loop <= this->NumPlantConnections; ++loop) {
+            if (calledFromLocation.loopNum != this->Loop(loop).LoopNum) continue;
+            if (calledFromLocation.loopSideNum != this->Loop(loop).LoopSideNum) continue;
+            thisLoop = loop;
+        }
```

Comment:
This "feels" like an odd way to do this, but I'm sure it's just pulled from the original code, so no need to change anything, just unusual. :man_shrugging: